src/expm_frechet.jl

```diff
@@ -0,0 +1,518 @@
+    Lw = A6 * Lw1 + M6 * W1 + Lw2
+    Lu = A * Lw + E * W
+    Lv = A6 * Lz1 + M6 * Z1 + Lz2
+    return U, V, Lu, Lv
+end
+
+@doc raw"""
+    expm_frechet(A, E)
+    expm_frechet!(buff, A, E)
+
+Compute Frechet derivative of expm(A) in direction E using algorithm 6.4 of [^AlMohyHigham2009]
+````math
+\mathrm{expm\_frechet}(A, E) = \lim_{t\to 0}\frac{1}{t}(\exp(A + tE) - \exp(A))
+````
+For sufficiently small ``\lVert A\rVert_1`` norm, we use Padé with an appropriate number of terms
+( one of (3, 5, 7, 9, 13)), with 13 terms is the default. Otherwise, we use the formula
+``\exp(A) = (\exp(A/2^s))^{2^s}``, where s is used to scale so ``2^{-s}\lVert A\rVert_1`` is smaller than element ``\mathrm{ell\_table\_61}``[14] of the lookup table ``\mathrm{ell\_table\_61}`` in the code, which comes from table 6.1 in op. cit.
+
+For ``\mathrm{expm\_frechet}!``, buff is a matrix of size 16*k times k, where ``k`` is the size of ``A``. The returns, (exp(A), dexp(A, E)) are stored in the first two blocks of buff.
+The remaining blocks are used as temporary storage. This is for algorithms calling expm_frechet repeatedly, where allocation becomes the biggest bottleneck.
+
+[^AlMohyHigham2009]:
+    >Al-Mohy, A. H., Higham, N. J. (2009)
+    >"Computing the Frechet Derivative of the Matrix Exponential, with an application to Condition Number Estimation."
+    >SIAM Journal On Matrix Analysis and Applications., 30 (4). pp. 1639-1657. ISSN 1095-7162
+    >doi: [https://doi.org/10.1137/080716426](https://doi.org/10.1137/080716426)
+"""
+function expm_frechet(A, E)
+    n = size(A, 1)
+    s = nothing
+    A_norm_1 = maximum(sum(abs.(A), dims=1))
+    m_pade_pairs = ((3, _diff_pade3), (5, _diff_pade5), (7, _diff_pade7), (9, _diff_pade9))
+
+    for m_pade in m_pade_pairs
+        m, pade = m_pade
+        if A_norm_1 <= ell_table_61[m]
+            U, V, Lu, Lv = pade(A, E)
+            s = 0
+            break
+        end
+    end
+    if isnothing(s)
+        # scaling
+        s = max(0, Int(ceil(log2(A_norm_1 / ell_table_61[14]))))
+        # pade order 13
+        U, V, Lu, Lv = _diff_pade13((2.0^-s) * A, (2.0^-s) * E)
+    end
+    # factor once and solve twice    
+    lu_piv = lu(-U + V)
+    eA = lu_piv \ (U + V)
+    eAf = lu_piv \ (Lu + Lv + (Lu - Lv) * eA)
+
+    # squaring
+    for k in 1:s
+        eAf = eA * eAf + eAf * eA
+        eA = eA * eA
+    end
+
+    return eA, eAf
+end
+function expm_frechet!(buff, A, E)
+    n = size(A, 1)
+    s = nothing
+    A_norm_1 = maximum(sum(abs.(A), dims=1))
+    k = size(A)[1]
+    m_pade_pairs =
+        ((3, _diff_pade3!), (5, _diff_pade5!), (7, _diff_pade7!), (9, _diff_pade9!))
+
+    for m_pade in m_pade_pairs
+        m, pade = m_pade
+        if A_norm_1 <= ell_table_61[m]
+            U, V, Lu, Lv = pade(buff, A, E)
+            s = 0
+            break
+        end
+    end
+    if isnothing(s)
+        # scaling
+        s = max(0, Int(ceil(log2(A_norm_1 / ell_table_61[14]))))
+        # pade order 13
```

Comment:
What is the goal of this comment if the next line calls a function of same name? And why are there so many other diff_pade13 functions? To more this file is a little chaotic and most of its content a little strange – sorry, but it is really mainly the problem that I will have to be able to maintain that and in its current form, this is just not possible.

Reply:
will clean up - and make diff_pade13 calling diff_pade13!